torch/distributed/distributed_c10d.py

```diff
@@ -1158,6 +1160,63 @@ def all_gather(tensor_list,
         work.wait()
 
 
+def all_gather_coalesced(output_tensor_lists,
+                         input_tensor_list,
+                         group=group.WORLD,
+                         async_op=False):
+    """
+    Gathers input tensors from the whole group in a list in a coalesced manner.
+
+    Arguments:
+        output_tensor_lists (list[list[Tensor]]): Output list. It should contain
+            correctly-sized tensors to be used for output of the collective.
+        input_tensor_list (list[Tensor]): Tensors to be broadcast from current process.
+        group (ProcessGroup, optional): The process group to work on
+        async_op (bool, optional): Whether this op should be an async op.
+            currently does not support async_op being True
+
+    Returns:
+        Async work handle, if async_op is set to True.
+        None, if not async_op or if not part of the group
+
+    """
+    _check_tensor_list(input_tensor_list, "tensor_list")
+    print("coalesce",  group)
+    if _rank_not_in_group(group):
+        print("not in group")
+        return
+
+    # Flatten the input and create a list of flat outputs.
+    input_coalesced = torch.cat([t.flatten() for t in input_tensor_list])
+    output_coalesced = [
+        torch.empty(input_coalesced.numel()) for _ in output_tensor_lists]
```

Comment:
This implicitly assumes size match. Let's make that explicit by adding size checks.

Reply:
Done